lib/semantics/check-omp-structure.cc

```diff
@@ -148,9 +164,45 @@ void OmpStructureChecker::Enter(const parser::OmpLoopDirective::Simd &) {
   SetContextAllowedOnce(allowedOnce);
 }
 
+// 2.8.3 do-simd-clause -> do-clause |
+//                         simd-clause
+void OmpStructureChecker::Enter(const parser::OmpLoopDirective::DoSimd &) {
+  SetContextDirectiveEnum(OmpDirective::DO_SIMD);
+
+  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::FIRSTPRIVATE,
+      OmpClause::LASTPRIVATE, OmpClause::LINEAR, OmpClause::REDUCTION,
+      OmpClause::ALIGNED};
+  SetContextAllowed(allowed);
+
+  OmpClauseSet allowedOnce{OmpClause::SCHEDULE, OmpClause::COLLAPSE,
+      OmpClause::ORDERED, OmpClause::SAFELEN, OmpClause::SIMDLEN};
+  SetContextAllowedOnce(allowedOnce);
+}
+
+// 2.11.4 parallel-do-simd-clause -> parallel-clause |
+//                                   do-simd-clause
+void OmpStructureChecker::Enter(
+    const parser::OmpLoopDirective::ParallelDoSimd &) {
+  SetContextDirectiveEnum(OmpDirective::PARALLEL_DO_SIMD);
+
+  OmpClauseSet allowed{OmpClause::DEFAULT, OmpClause::PRIVATE,
+      OmpClause::FIRSTPRIVATE, OmpClause::SHARED, OmpClause::COPYIN,
+      OmpClause::REDUCTION, OmpClause::LASTPRIVATE, OmpClause::LINEAR,
+      OmpClause::ALIGNED};
+  SetContextAllowed(allowed);
+
+  OmpClauseSet allowedOnce{OmpClause::IF, OmpClause::NUM_THREADS,
+      OmpClause::PROC_BIND, OmpClause::SCHEDULE, OmpClause::COLLAPSE,
+      OmpClause::ORDERED, OmpClause::SAFELEN, OmpClause::SIMDLEN};
+  SetContextAllowedOnce(allowedOnce);
+}
+
 void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
   // 2.7 Loop Construct Restriction
-  if (GetContext().directive == OmpDirective::DO) {
+  if (GetContext().directive == OmpDirective::DO ||
```

Comment:
I wonder if it would be better (cleaner to read) to use a `switch` statement here instead of `if` here?
Even I'm not sure, anyone else have opinions on this?
```suggestion
  switch(GetContext().directve) {
  case OmpDirective::DO:
  case OmpDirective::PARALLEL_DO:
  case OmpDirective::DO_SIMD:
  case OmpDirective::PARALLEL_DO_SIMD:
    <code block>
    break;

  case OmpDirective::SIMD:
  case OmpDirective::DO_SIMD:
  case OmpDirective::PARALLEL_DO_SIMD:
    <code block>
    break;
  }
```

Reply:
I think that the `switch` is more clear.

Reply:
Updated with `EnumSet` instead.